### PR TITLE
kata-agent: add network gRPC support

### PR DESCRIPTION
**Pull request by egernst**
Introduce support for setting up interface and routes when working with
Kata agent.

Signed-off-by: Eric Ernst <eric.ernst@intel.com>

### CODE REVIEW

**Comment by egernst** [comment]
Functioning network gRPC calls dependent  on https://github.com/kata-containers/agent/pull/126

**Comment by egernst** [comment]
This is dependent on the gRPC changes landing in kata-containers/agent and being re-vendorded.

File: kata_agent.go [hunk 1]
```diff
@@ -295,6 +295,82 @@ func (k *kataAgent) exec(pod *Pod, c Container, cmd Cmd) (*Process, error) {
...
+	networkNS, err := pod.storage.fetchPodNetwork(pod.id)
+	if err != nil {
+		return err
+	}
+
+	if networkNS.NetNsPath == "" {
+		return nil
+	}
+
+	for _, endpoint := range networkNS.Endpoints {
```
Comment (jodh-intel) [inline kata_agent.go]: Random thought: this code is pretty similar to:

- https://github.com/containers/virtcontainers/blob/master/hyperstart_agent.go#L173

I wonder if it would be worth refactoring to make a generic function in `agent.go` that accepted a `Pod` and a function to operate on the `Endpoint`'s?

File: kata_agent.go [hunk 2]
```diff
@@ -295,6 +295,82 @@ func (k *kataAgent) exec(pod *Pod, c Container, cmd Cmd) (*Process, error) {
...
+		var route grpc.Route
+		var ipAddresses []*grpc.IpAddress
+		for _, addr := range endpoint.Properties().Addrs {
+			// Skip IPv6 because not supported
```
[the remark this answers is not in the document]
Comment (egernst) [reply kata_agent.go]: Yes: https://github.com/containers/virtcontainers/issues/579

File: kata_agent.go (hunk 1 above)
Comment (egernst) [reply kata_agent.go]: It is similar in that it loops over the endpoints and addresses, but with the protocols being different the implementation is different enough (operating on an interface and route at a time) that I think it may be a bit painful for a potentially short term code reuse benefit.  WDYT?

File: kata_agent.go [hunk 3]
```diff
@@ -330,6 +406,10 @@ func (k *kataAgent) startPod(pod Pod) error {
 		hostname = hostname[:maxHostnameLen]
 	}
 
+	if err := k.sendNetworkInterfaceAndRouteRequests(pod); err != nil {
```
[the remark this answers is not in the document]
Comment (egernst) [reply kata_agent.go]: Hey @jodh-intel - I started to do this a couple of times and am having a hard time.  The current "mocking" in kata_agent_test.go is pretty limited (return nil, nil), so it is pretty hard to exercise without re-implementing a lot of logic from the existing kata agent in the test file.  I now recall all of the mock discussions that you were driving back in the CC-3.0 pre-release time frame.  

Let me think on how to handle this pragmatically.  Let me know if you have suggestion in the meantime.

**Comment by egernst** [comment]
Tested with latest agent and this PR now includes revendoring for latest gRPC changes.  @sboeuf @jodh-intel  can you take a look?  Would be good to get working networking in...

**Comment by sboeuf** [comment]
@egernst I'll review it today !

File: kata_agent.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jodh-intel) [reply kata_agent.go]: I'm not sure I understand. We don't necessarily need 100% coverage but this PR adds a relatively big function with no tests. I don't think we should be attempting to land significant features without tests.

The approach I'd take is to break `sendNetworkInterfaceAndRouteRequests()` into smaller functions (for improved clarity, code complexity and testing) and you should then be able to unit-test most of those. But even as it is, can't you create test functions to atleast handle?:

- `fetchPodNetwork()` failing.
- `networkNS.NetNsPath == ""`.
- `addr.IP.IsLoopback() == true` (force it with `link, err := netlink.LinkByName("lo")` ?)

**Comment by jodh-intel** [comment]
Yep - would be great to get networking in, once we have tests for it.

File: kata_agent.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (egernst) [reply kata_agent.go]: Yes, you are right. I will refactor and add tests today.

File: kata_agent.go [hunk 4]
```diff
@@ -335,6 +410,45 @@ func (k *kataAgent) startPod(pod Pod) error {
 		hostname = hostname[:maxHostnameLen]
 	}
 
+	//
+	// Setup network interfaces and routes
+	//
+	networkNS, err := pod.storage.fetchPodNetwork(pod.id)
```
Comment (sboeuf) [inline kata_agent.go]: Note: This should be part of the `pod` and we should be able to access those infos from in-memory. Opened an issue here: https://github.com/containers/virtcontainers/issues/628
Will be fixed in another PR

File: kata_agent.go [hunk 5]
```diff
@@ -335,6 +410,45 @@ func (k *kataAgent) startPod(pod Pod) error {
...
+	if err != nil {
+		return err
+	}
+	interfaces, routes, err := k.generateKataInterfacesAndRoutes(networkNS)
+	if err != nil {
+		return err
+	}
```
Comment (sboeuf) [inline kata_agent.go]: Please add a blank line to let the code breathe :smile:
And globally, this comment applies to every place in the code where there are not enough blank lines (for readability)

File: kata_agent.go [hunk 6]
```diff
@@ -335,6 +410,45 @@ func (k *kataAgent) startPod(pod Pod) error {
...
+		ifcReq := &grpc.UpdateInterfaceRequest{
+			Interface: ifc,
+		}
+		resultingInterface, err := k.sendReq(ifcReq)
+		if err != nil {
+			k.Logger().WithFields(logrus.Fields{
+				"interface requested": fmt.Sprintf("%+v", ifc),
+				"resulting interface": fmt.Sprintf("%+v", resultingInterface),
+			}).WithError(err).Error("update interface request failed")
+			continue
```
Comment (sboeuf) [inline kata_agent.go]: Why don't we fail here ? Have we agreed on ignoring failures here ?

File: kata_agent.go [hunk 7]
```diff
@@ -335,6 +410,45 @@ func (k *kataAgent) startPod(pod Pod) error {
...
+		//send add route request
+		routeReq := &grpc.AddRouteRequest{
+			Route: route,
+		}
+		_, err = k.sendReq(routeReq)
+		if err != nil {
+			k.Logger().WithFields(logrus.Fields{
+				"route requested": fmt.Sprintf("%+v", route),
+			}).WithError(err).Error("update route request failed")
+			continue
```
Comment (sboeuf) [inline kata_agent.go]: Why don't we fail here ? Have we agreed on ignoring failures here ?

File: kata_agent.go [hunk 8]
```diff
@@ -300,6 +300,81 @@ func (k *kataAgent) exec(pod *Pod, c Container, cmd Cmd) (*Process, error) {
...
+			}
+			netMask, _ := addr.Mask.Size()
+			ipAddress := grpc.IPAddress{
+				Family:  grpc.IPFamily_v4,
+				Address: addr.IP.String(),
+				Mask:    fmt.Sprintf("%d", netMask),
+			}
+			ipAddresses = append(ipAddresses, &ipAddress)
+		}
+		ifc := grpc.Interface{
```
Comment (sboeuf) [inline kata_agent.go]: You can do
```go
ifc := &grpc.Interface{
```
so that later in the code you can do
```go
ifaces = append(ifaces, ifc)
```

File: kata_agent.go [hunk 9]
```diff
@@ -300,6 +300,81 @@ func (k *kataAgent) exec(pod *Pod, c Container, cmd Cmd) (*Process, error) {
...
+			IPAddresses: ipAddresses,
+			Device:      endpoint.Name(),
+			Name:        endpoint.Name(),
```
[the remark this answers is not in the document]
Comment (egernst) [reply kata_agent.go]: Done: https://github.com/kata-containers/agent/issues/147

File: kata_agent.go (hunk 4 above)
Comment (egernst) [reply kata_agent.go]: ack

File: kata_agent.go (hunk 5 above)
Comment (egernst) [reply kata_agent.go]: Will add some air.

File: kata_agent.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sboeuf) [reply kata_agent.go]: Thanks !

File: kata_agent.go (hunk 6 above)
Comment (egernst) [reply kata_agent.go]: I agree on logging the error and continuing to apply remaining interfaces and routes.  I think that this is an error, but not necessarily catastrophic -- user can get into the workload and then assess what failed in setup.  WDYT?

File: kata_agent.go (hunk 8 above)
Comment (egernst) [reply kata_agent.go]: yawn.  Sure. :)

File: kata_agent.go (hunk 6 above)
Comment (sboeuf) [reply kata_agent.go]: Well, only logging without returning this error might be confusing since the user will expect everything to work as expected if he does not check the logs. And he might not understand why he does not get what he expects inside the container.
We could have a retry mechanism (maybe retry 5 times with `100ms` between each time), so that we really make sure there is a real issue here, that would worth throwing an error.

File: kata_agent.go (hunk 4 above)
Comment (egernst) [reply kata_agent.go]: done.

**Comment by egernst** [comment]
@sboeuf -- updated.

File: kata_agent.go (hunk 6 above)
Comment (sboeuf) [reply kata_agent.go]: @sameo @jodh-intel WDYT ?

Comment (sameo) [reply kata_agent.go]: @egernst @sboeuf If an interface or a route failed to be set, it's likely that networking will be at least partly dysfunctional, leaving the end-user with a potentially non working pod. By only logging this and not failing we'd give a broken pod to the upper layers of the stack and pretend it's all fine, which is bad from a end user experience perspective. In other words, we should log **and** fail here and the pod must not be passed upward. It should be stopped at least.

Comment (jodh-intel) [reply kata_agent.go]: +1 for failing - if we continue, surely the system is in an unknown state (bad)? Turning it round, do we ever think this might fail? If so, why?

File: kata_agent.go [hunk 10]
```diff
@@ -300,6 +300,81 @@ func (k *kataAgent) exec(pod *Pod, c Container, cmd Cmd) (*Process, error) {
...
+					k.Logger().WithFields(logrus.Fields{
+						"unsupported-route-type": "ipv6",
+						"destination":            route.Dest,
+					}).Warn("unsupported route")
+					continue
+				}
+			}
+
+			route.Gateway = r.Gw.String()
+			if route.Gateway == "<nil>" {
```
[the remark this answers is not in the document]
Comment (egernst) [reply kata_agent.go]: Agreed

File: kata_agent.go (hunk 6 above)
Comment (egernst) [reply kata_agent.go]: Agreed. Thanks guys.

Comment (egernst) [reply kata_agent.go]: I should have stopped with error in the first place, mea culpa.  It won't work today as I didn't handle conflicting routes (for the initial interface route in the guest).    I have a new PR, https://github.com/kata-containers/agent/pull/148, R on the agent that I started (not ready just yet) which will act on a set of routes instead, acting fully declarative.  Without this I don't think we can really test networking in a container without bypassing conflict errors (because the initial interface route already exists).  This'll simplify our handling quite a bit --   I'll try to get this ready for merge today.

**Comment by egernst** [comment]
see https://github.com/kata-containers/agent/pull/148

File: kata_agent.go [hunk 11]
```diff
@@ -300,6 +300,81 @@ func (k *kataAgent) exec(pod *Pod, c Container, cmd Cmd) (*Process, error) {
 	return prepareAndStartShim(pod, k.shim, c.id, req.ExecId, k.state.URL, cmd)
 }
 
+func (k *kataAgent) generateKataInterfacesAndRoutes(networkNS NetworkNamespace) ([]*grpc.Interface, []*grpc.Route, error) {
```
[the remark this answers is not in the document]
Comment (egernst) [reply kata_agent.go]: apparently I really like to push the kata brand, in this case in error.  Will update...

**Comment by egernst** [comment]
waiting for kata-containers/agent#148 to land so I can update the revendoring.

In the meantime, addressed feedback from @sameo and @jodh-intel and @sboeuf .

**Comment by egernst** [comment]
@sboeuf  PTAL?

File: kata_agent.go (hunk 4 above)
Comment (sboeuf) [reply kata_agent.go]: You can update this cause you don't need to fetch the pod network anymore since #628 has been solved by #629

**Comment by egernst** [comment]
@jodh-intel @sboeuf  - updated based on updated pod structure.  PTAL

**Comment by sboeuf** [comment]
LGTM

File: kata_agent.go [hunk 12]
```diff
@@ -335,6 +412,45 @@ func (k *kataAgent) startPod(pod Pod) error {
...
+	}
+	for _, ifc := range interfaces {
+		// send update interface request
+		ifcReq := &grpc.UpdateInterfaceRequest{
+			Interface: ifc,
+		}
+		resultingInterface, err := k.sendReq(ifcReq)
+		if err != nil {
+			k.Logger().WithFields(logrus.Fields{
+				"interface requested": fmt.Sprintf("%+v", ifc),
```
Comment (jodh-intel) [inline kata_agent.go]: Nit: there is an informal understanding [*] that all log fields should be single-word (so I'd just replace spaces with hyphens here and below).

---

[*] - yep, we should make expectations formal by writing a doc! ;)

**Comment by sameo** [comment]
LGTM

**Comment by sameo** [comment]
@sboeuf @egernst Thanks for the PR. Let's merge it once CI passes.
